### ISSUE TITLE
NV-14 Phanton Telemetry RPM from FC's

### ISSUE THREAD

**Issue by Outwide** [EdgeTX/edgetx#2721]
### Is there an existing issue for this problem?

- [X] I have searched the existing issues

### What part of EdgeTX is the focus of this bug?

Transmitter firmware

### Current Behavior

Phantom RPM telemetry showing in NV-14 display telemetry display. https://github.com/EdgeTX/edgetx/labels/bug%20%F0%9F%AA%B2
Telemetry display when using BF4.3 Fr 10.8 shows one RPM set at 1000
Telemetry display when using Inav 5.1 shows two RPM set at 0 and the other counting up

![BF3 4 PH RPM](https://user-images.githubusercontent.com/89946386/201496460-c83d3f9d-d154-48a4-b3f0-59727514ffe5.jpg)
![Inav5-1 PH RPM](https://user-images.githubusercontent.com/89946386/201496465-574416e6-ab1b-49d4-9e3b-34ff24458d22.jpg)

### Expected Behavior

Only show actual telemetry transmitted, No phantoms. https://github.com/EdgeTX/edgetx/labels/bug%20%F0%9F%AA%B2

### Steps To Reproduce

I have setup test rigs for both HGLRC Zeus F722 and Matek H743 Slim both FC's can be used with either BF 4.3 Fr10.8 and Inav 5.1.

Clean install of BF 4.3 Fr10.8 both FC's (HGLRC Zeus F722 and Matek H743 Slim) shows one phantom telemetry RPM which is wrong. No motor or rpm sensor is attached.

Clean install of Inav5.1 both FC's (HGLRC Zeus F722 and Matek H743 Slim) show two phantom telemetry RPM which is wrong. No motor or rpm sensor is attached.

With Matek H743 Slim I tried it with motor and without but go the same result.

### Version

Other (Please specify below)

### Transmitter

Flysky NV14

### Anything else?

I like Edge TX you guys are doing fantastic work.
2.8.0 RC4

**Comment by Outwide** [comment]
1.  Which receiver are you using

  *   FS-IA6B
  *   FS_IA10B
  *   Note: I will also check with the, Crossover receiver XR602T-A and ELRS which is coming soon and let you know what I found.
2- I assume you sending FC telemetry to the iBUS input of the receiver, correct?

  *   Yes from the FC to the receiver and all working. You can see the wires at the bottom that connect to the receiver and I-Bus sensor. Picture attached

  1.  If you remove all sensors and run rediscover - doe the problem persist?

  *   Yes I have done rediscover about 10 times and have been trying to find this problem for months.

  1.  A3 telemetry - is it coming from the FC or from the FlySky sensor?

  *   The FC . No fly sensor or esc is attached. The A3 appears to be working fine and is the battery voltage no need to investigate this for me.

Just so you know I have been trying to get and RPM sensor working from the FC for months but it has appeared that I have been getting Phantom telemetry. The picture above is as you see it, clean install of operating programme with no ESC’s or anything else attached.

Note: The Flysky sensors work and I think with no problem. These are value sensor type.

I wish to use all sensors from the FC, which are a telemetry type sensors,  and add other telemetry to the FC and then out to EdgeTX.


Thank you for your reply and if I need be clearer let me know.

I like Edge TX

**Comment by eshifri** [comment]
It may be completely irrelevant, but on the second image there are two A3 sensors showing different voltage.
Is it correct behavior?

I have never tried any RPM sensor, so there is a chance RPM packets decoding never worked for NV14.

**Comment by Outwide** [comment]
Yes I am aware of A3, I did change the voltage scale and didn’t reset. Completely irrelevant.

On the picture I supplied of the 2 FC no esc or telemetry sensor is attached. I get the same phantom RPM telemetry on the 2 other Matek H743 slim FC’s which are connected to esc and other devices.

I have supplied a New picture of the test rig BF4.3 10.8 and just rechecked it again and still get phantom RPM telemetry and supposably run at a 1000RPM.

I will test the other receiver soon, Crossover receiver XR602T-A and then on the ELRS when it arrives.

Note: it maybe a receiver base problem

Thanks

**Comment by Outwide** [comment]
I went back to check Inav 5.1 and made sure things were turned off as per program. See pictures

Thanks


________________________________

**Comment by eshifri** [comment]
Sorry, I do not see any new pictures in this thread - only two original pictures.

**Comment by Outwide** [comment]
Sorry they just didn’t send

**Comment by pfeerick** [comment]
Images and files attached to emails are rejected by github - you'll need to upload them on the issue page (i.e. drag and drop the images to the text entry area to upload).

**Comment by Outwide** [comment]
Sorry, these are the images as per discussion above.

![edgeTX Inav 5 1](https://user-images.githubusercontent.com/89946386/202846036-17c0258d-024a-4ccc-a2f1-761325b610f8.jpg)
![Test rigs](https://user-images.githubusercontent.com/89946386/202846068-028fe6ef-c0c4-4489-98db-72d8f8f29d9a.jpg)
![Test rig setup](https://user-images.githubusercontent.com/89946386/202846072-f89448ed-ce51-49ee-8a1f-a28cc8abfbcd.jpg)
![Inav5 1](https://user-images.githubusercontent.com/89946386/202846076-3a4b3261-1ee3-4215-8682-cb69ba201ab1.jpg)

**Comment by eshifri** [comment]
Looks like you have old internal radio module FW (Sgnl==10 on your images).
I have no idea whether it will help, but you may want to try an updated version from here:
https://www.rcgroups.com/forums/showpost.php?p=45357535
Some discussions may be found here: https://www.rcgroups.com/forums/showthread.php?3023832-UndergroundFPV-Dark-Knight-Radio/page251

Telemetry processing is handled differently with the new FW (more or less on the MPM route).

**Comment by Outwide** [comment]
Hi thanks for the info. I ask how do I flash the "internal chip" FW and keep all the Edge TX info and working. I saw something about a special loader for that prepose, said it was under development.  
I just want to be very careful ! as I am not sure.

![FW version](https://user-images.githubusercontent.com/89946386/202869860-fc93c273-58c0-4753-9166-ca65fc0c08e9.jpg)

**Comment by eshifri** [comment]
There is an unpleasant surprise: this screen shows you have the latest version.
Unfortunately, there are two different radio module FWs in the field with the same version number.

In order to update the radio FW please be advised to put it into the FIRMWARE folder on the SD card (unzip it first, of course).
On the TX navigate to this folder and select the file. In the pop-up menu there will be an option to update the radio FW.

There is also one more thing we can try, maybe before that. I've got some ideas, but I cannot test them. Do you think you will be able to give a try for the ETX version I upload?

**Comment by Outwide** [comment]
Yes

**Comment by Outwide** [comment]
"Yes" Please I would like to try whatever method you recommend.

**Comment by eshifri** [comment]
[firmware.zip](https://github.com/EdgeTX/edgetx/files/10047933/firmware.zip)
Can you please give it a try?

**Comment by Outwide** [comment]
Ok I have put the new firmware in to the FIRMWARE BIN on the SD card and I will open the TX companion load it as stated above.

**Comment by Outwide** [comment]
Problem,  The NV-14 connected to the computer, using USB, see on screen on the Transmitter. In TX Companion I navigate to the SD card on the handset and FIRMWARE BIN and load the FIRMWARE file you sent. This was my procedure.  
![can't write](https://user-images.githubusercontent.com/89946386/202871513-d09d155c-2f43-4b8b-b05a-25225dd3a0c1.jpg)

**Comment by eshifri** [comment]
This is just an ETX FW. The fastest way to install it is from SD card: just copy it into the FIRMWARE folder, boot the TX in the bootloader mode (press both trims towards the center and switch the TX on). You will be able to install a new FW.

**Comment by eshifri** [comment]
Actually, I'm not sure Companion can install the FW that is on the SD card - never tried this.

**Comment by eshifri** [comment]
And one more version to try.
[firmware.zip](https://github.com/EdgeTX/edgetx/files/10047958/firmware.zip)

**Comment by Outwide** [comment]
Ok If I go into Bootloader mode I am not able to see the SD card.
I have tried moving both Blue button trims toward the centre of the screen and setting to Bootloader mode but I can't seed the SD card on the computer or tell TX to go get it from the bin.

Can I just tell TX to go to my computer and load the firmware.

**Comment by eshifri** [comment]
There are many ways to do it, but the fastest is:
1. Switch TX on, Connect it to the computer with USB, copy firmware.bin into the FIRMWARE folder.
2. Disconnect the USB and switch the TX off.
3. Switch TX on in the bootloader mode. There will be and option to "Write the firmware". 
4. All FWs on the SD card will be listed.
5. Follow the on-screen instructions.

Please start with the latest  version I have uploaded.

**Comment by Outwide** [comment]
3. Switch TX on in the bootloader mode. There will be and option to "Write the firmware".
When I do this it is jut a black screen. Not comes up you can only to your computer

OK I can start the Transmitter and navigate to the Radio setting Tab and to Firmware bin and load the select the firmware and flash.
 
![Firmware](https://user-images.githubusercontent.com/89946386/202872574-60687728-825c-4005-a0f9-630784d4c8e7.jpg)

**Comment by Outwide** [comment]
![worked](https://user-images.githubusercontent.com/89946386/202872763-8fdfbd64-ccd6-429e-b70d-c9b302a9722f.jpg)

**Comment by Outwide** [comment]
Ok I have deleted the old telemetry screen and new discover. 
![NEW Telemetry screen](https://user-images.githubusercontent.com/89946386/202872892-fc26b1ce-3a9d-4f36-b7bd-a1af00959f7d.jpg)

**Comment by eshifri** [comment]
Unfortunately, this is flashing a bootloader only.
To flash the FW itself you need to be in the bootloader mode.

Let's check you are doing it correctly: press both trims towards the center and both power buttons simultaneously.
You should see the bootloader menu.

OK, this is from my latest build?  
I do not know what to make from this: there are two RPM sensors in the FlySky system. So it might be possible that one returns zero and the other - random number.

**Comment by Outwide** [comment]
Yes it worked this time. RPMs are a zero both.
![RPM at Zero](https://user-images.githubusercontent.com/89946386/202873435-f22a42b2-00d0-44e4-9002-4f6c62dc6fe3.jpg)

**Comment by eshifri** [comment]
Aha! Some progress! Thank you for testing this!
Do you have any way to test it with the motor? To see measured values?

**Comment by Outwide** [comment]
I would have to say No because as Inav uses the belhli32 to supply RPM signal, for the smaller ESC, to the OSD only not out to the Transmitter. I have asked for it to be considered on the Inav github but as I am a no one so it may not be heard. I am wanting to get a 2 wire system such as the hobby wing RPM sensor, which from my investigation, only outputs a signal of difference which then has to be coded in the FC so it can come out to the OSD and Transmitter. This what I am ask help for, someone to do some coding, as it is very important for preflight checks and documentation for drone licensing, as if there is something wrong you will see it progress.
Note: I have contacted CoderElectronics to see if he can help me with some code you may wish to touch base with him. All help is good.

Note: I only have the original Flysky RPM sensor which I had no problem with but not suitable with FC

**Comment by Outwide** [comment]
I personal say Thank you for your generous help. You are truly helping me and other so much.

**Comment by eshifri** [comment]
You are most welcome. 

Are you saying FlySky RPM sensor worked properly without my changes?

**Comment by Outwide** [comment]
I can't say that yet as I don't have a rig setup for that yet. My understanding is that FlySky RPM sensor is a value sensor and FC's transmit "telemetry". Tricky wording which mean you can't connect them both together as one cancels out the other.

I am tiring to get all telemetry out of an FC and I am nearly there. Just the 2 wire system such as the hobby wing RPM sensor left.

**Comment by Outwide** [comment]
I would say that the issue can be closed for now as both RPM's show zero.

**Comment by eshifri** [comment]
Again, I have never tried, but all FlySky sensors are digital and communicate with RX through iBUS (Serial port). They can be daisy-chained. You can have several sensors reporting the same value - they will be assigned different IDs and will be visible.

**Comment by Outwide** [comment]
Yes I have tried that. But if you are wanting a small footprint in your aircraft an FC is a good way to go as you get a lot of telemetry output.

**Comment by Outwide** [comment]
Hi I ask would it be ok to keep in contact with you. I am on Discord it would be nice to have some friends I can talk with.

**Comment by eshifri** [comment]
Sure!

**Comment by eshifri** [comment]
Actually, it should not be closed since the code was not committed yet.

**Comment by pfeerick** [comment]
Reopened

> Actually, I'm not sure Companion can install the FW that is on the SD card - never tried this.

Probably not - since you really should be in DFU mode when doing that (so the SD card will not be visible). Online flashing when the SD card *is* visible is just begging for disaster. Thus flashing via the bootloader on the transmitter itself is the simplest, safest and fastest... 

Just for reference, with 2.8.0-RC4 firmware, I get the following when the NV14 and FS-iA6B have no sensors connected, 

![image](https://user-images.githubusercontent.com/5500713/202876633-152e85eb-03c8-431d-8336-e0eec3d3f21a.png)

If I have either the CPD01 (Magnetic) or CPD02 (Optical) Sensors I get only one extra sensor entry:

![image](https://user-images.githubusercontent.com/5500713/202876661-5d782de7-eb92-4c00-bfed-5a437d82cb69.png)

If I have both connected, I get two entries, with different instances:

![image](https://user-images.githubusercontent.com/5500713/202876676-e3e2562b-ba5e-4168-b592-674b6b63f781.png)

Finally both the CPDO1 and CPD02 working... they pretty much agree on the speed of the motor, which was nice. 

![image](https://user-images.githubusercontent.com/5500713/202876750-a4e65a2f-689d-4d16-8df0-cc826854657a.png)

**Comment by eshifri** [comment]
Hm, then it should be probably closed "with no problem found".
I'm a little bit surprised it works as is, since I thought I have found a problem. :-)
If it works with native sensors the problem must be on the FC side, right? 
The only difference I see here is you are using the updated radio module FW (Sgnl==99).
The telemetry processing is different for old and new versions:
https://github.com/EdgeTX/edgetx/blob/f84c26b6152b3cd1a497bea49eca4fbc95b5e1c1/radio/src/telemetry/flysky_nv14.cpp#L272
My change, most likely, affected the older FW route.
Unfortunately, I do not have these sensors and cannot test.
Shall I submit a PR? You will be able to see whether it breaks something. :-)

**Comment by pfeerick** [comment]
Sure, can at least give it a try and let you know. :) 

Yes, I updated the NV14 internal RF back when you first told me there was a newer FW for it... and that really does need to be the minimum standard here since it exists and fixes things. So if it only affects older FW, users of that are on their own unless they complain 🤷‍♂️ . 

Until a comparison is done to show telemetry with the FC connected, and telemetry without the FC connected (after sensor deletion and rediscovery, of course) which proves otherwise, I would strongly be leaning towards something in the Betaflight/iNav telemetry implementation for iBus is a bit broken, thus that time tested IT adage of "💩 in, 💩 out" would be applicable here.

**Comment by Outwide** [comment]
Hi, again thank you both so much for your help. So I ask, what should I do? get the new Main board V1.6. Will you guys let the Dev's in Inav 5.1 know there is an issue in their software as a common curtsey. 
Just so you know I still have to test another two receivers connect to an FC to see what I get. and then I can reconnect back to the old style "Box receiver "CPD02 (RPM), Temp and Voltage.

**Comment by eshifri** [comment]
I do not think  you should change the main board unless you have other reasons to do it.
You MAY want to upgrade the radio module FW (my links above).
I have submitted a PR and would suggest you wait for some testing.
Since you do have the FlySky sensors you may want to check how they work with ETX before and after the fix.
Do not rush. :-)

**Comment by Outwide** [comment]
I see it has finished checking do I need to get this file and how or just wait.

**Comment by eshifri** [comment]
It is exactly the same binary file I gave you above.

**Comment by pfeerick** [comment]
Specifically, as mentioned in this comment https://github.com/EdgeTX/edgetx/issues/2721#issuecomment-1320950171. You copy it onto your radio, put it in a folder (I suggest FIRMWARE also for this), and then with the radio running normaly, go into radio settings, SD, FIRMWARE and select the file - you'll get an option to flash the internal rf module.

**Comment by Outwide** [comment]
Hi eshifri how do I contact you as I couldn't see you on Discord.
I have tried this receiver, Crossover receiver XR602T-A which has telemetry on the HGLRC Zeus F722, using Inav 5.1 and was not able to get it to work in Inav, was able to get it to bind. I also tried in with Beta flight 4.3 and was able to get it to work in Beta flight and see stick movement in the receiver section but kept getting RF signal warning critical. I have noticed at the top right-hand side, signal bars where all black. I will try with my other TX tomorrow which doesn't have the update on yet and if there is a change, I will let you know. ExpressLRS does not appear to be ready yet form the NV-14. It would appear that there are few receivers with telemetry for Flysky AFHDS 2A and AFHDS 3 is still broken.

**Comment by eshifri** [comment]
Hi @Outwide I'm at the Discord under the same name. You can find me in ETX forum (usually, in the dev's section).
I think ELRS works on NV14, but again: I have never tried. :-)
RF signal warning may be misleading - it is not standardized over different protocols and it is possible the warning levels are incorrect for your protocol. Please take a look at the signal quality numbers in the telemetry (if you receive it).

**Comment by pfeerick** [comment]
ELRS/CRSF are both working fine for me with the NV14 in both 2.7.1 and 2.8.0. 

I don't know anything about Crossover XR602T-A other than it exists, and seems to use AFHDS2A, but the Flysky iA6B and iA10B and sensors are working fine as well. I had to update the firmware on the Flysky iA6B though (that was fun!) as I found it had a problem with channel outputs with the original firmware it shipped with (was a really old version). That's the only major issue I've had (other than the NV14 internal RF needing to be updated also to fix bugs like the signal quality value).
